### PR TITLE
CORE-9662: Make SignatureSpec an interface on public API

### PR DESCRIPTION
**Pull request by anien**
* adds `SignatureSpec` implementation and updates use sites
* adds `ParameterizedSignatureSpec` (moved from api repo)

API PR: [#1003](https://github.com/corda/corda-api/pull/1003)

### CODE REVIEW

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+//    constructor(signatureName: String, customDigestName: DigestAlgorithmName)
+//            : this(signatureName, customDigestName, null)
+
+
+    companion object {
+        /**
+         * SHA256withRSA [SignatureSpec].
+         */
+        val RSA_SHA256 = SignatureSpecImpl("SHA256withRSA")
```
Comment (kyriathar) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt]: Would extracting the signature specs into a file/class called `SignatureSpecs` look better?

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    /**
+     * The signature-scheme name as required to create [java.security.Signature] objects
+     * (e.g. `SHA256withECDSA`).
+     */
+
+//    fun SignatureSpec(signatureName: String) {
+//        requireNotNull(signatureName) { "The signatureName must not be null." }
+//        require(!signatureName.isBlank()) { "The signatureName must not be blank." }
+//        this.signatureName = signatureName
+//    }
```
Comment (kyriathar) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt]: We need to enable those checks for `signatureName` on construction.

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+//    constructor(signatureName: String, params: AlgorithmParameterSpec)
+//            : this(signatureName, null)
+
+    /**
+     * Converts a [ParameterizedSignatureSpec] object to a string representation.
+     *
+     * @return string representation
+     */
+    override fun toString(): String {
+        return if(signatureName != null) {
```
Comment (kyriathar) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt]: I think instead let's make `signatureName` not nullable then these checks can be removed

Comment (anien) [reply libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt]: Done.

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt (hunk 2 above)
Comment (anien) [reply libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt]: I believe `requireNotNull` should not be needed anymore. I have put back `isBlank` check.

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * @param signatureName A signature-scheme name as required to create [java.security.Signature]
+ *                      objects (for example, `SHA256withECDSA`).
+ * @param params        Signature parameters. For example, if using `RSASSA-PSS`, to avoid
+ *                      using the default SHA1, you must specify the signature parameters explicitly.
+ *
+ *                      When used for signing, the `signatureName` must match the corresponding key scheme,
+ *                      for example, you cannot use `SHA256withECDSA` with `RSA` keys.
+`` */
+class ParameterizedSignatureSpec(private val signatureName: String, val params: AlgorithmParameterSpec) : SignatureSpec {
```
Comment (blsemo) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt:18]: Should this not be derived from `SignatureSpecImpl`? Otherwise it will have to duplicate the implementation.

File: libs/crypto/cipher-suite/src/test/java/net/corda/crypto/cipher/suite/SignatureSpecEqualityJavaApiTest.java
Comment (blsemo) [inline libs/crypto/cipher-suite/src/test/java/net/corda/crypto/cipher/suite/SignatureSpecEqualityJavaApiTest.java:1]: Why did you move Java API tests? Nothing outside of our control can call code in this repo from Java, so they should not be here, or if they test anything else but callability from Java, they should be renamed.

File: libs/crypto/cipher-suite/src/test/kotlin/net/corda/crypto/cipher/suite/SignatureSpecEqualityTests.kt [hunk 5]
```diff
@@ -12,12 +11,11 @@ import kotlin.test.assertTrue
...
-        private val SIGNATURE_SPEC_10 = SignatureSpec("SHA256withRSA")
-        private val SIGNATURE_SPEC_100 = SignatureSpec("sha256WITHrsa")
-        private val SIGNATURE_SPEC_2 = SignatureSpec("SHA384withRSA")
-        private val PARAMETERIZED_SIGNATURE_SPEC_1 = ParameterizedSignatureSpec(
-            "RSASSA-PSS",
+        private val SIGNATURE_SPEC_1 = SignatureSpecImpl("SHA256withRSA")
+        private val SIGNATURE_SPEC_10 = SignatureSpecImpl("SHA256withRSA")
+        private val SIGNATURE_SPEC_100 = SignatureSpecImpl("sha256WITHrsa")
+        private val SIGNATURE_SPEC_2 = SignatureSpecImpl("SHA384withRSA")
+        private val PARAMETERIZED_SIGNATURE_SPEC_1 = ParameterizedSignatureSpec("",
```
Comment (blsemo) [inline libs/crypto/cipher-suite/src/test/kotlin/net/corda/crypto/cipher/suite/SignatureSpecEqualityTests.kt]: Why woulld a `ParameterizedSignatureSpec` work with an empty name, but not a blank (all whitespace) name (see file above)?

File: notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt [hunk 6]
```diff
@@ -70,16 +74,19 @@ fun validateRequestSignature(notarizationRequest: NotarizationRequest,
...
-    val signature = signingService.sign(
-        serializedRequest,
-        myLegalIdentity,
-        SignatureSpec.ECDSA_SHA256 // TODO This shouldn't be hardcoded?
-    )
+    val signature = signatureSpecService.defaultSignatureSpec(myLegalIdentity)?.let {
+        signingService.sign(
+           serializedRequest,
+            myLegalIdentity,
+            it // TODO This shouldn't be hardcoded?
```
Comment (blsemo) [inline notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt]: Isn't the todo mow obsolete?

Comment (anien) [reply notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt]: It actually is. Removing.

File: libs/crypto/cipher-suite/src/test/java/net/corda/crypto/cipher/suite/SignatureSpecEqualityJavaApiTest.java
Comment (anien) [reply libs/crypto/cipher-suite/src/test/java/net/corda/crypto/cipher/suite/SignatureSpecEqualityJavaApiTest.java:1]: I didn't move the test. I simply fix the test.

File: libs/crypto/cipher-suite/src/test/kotlin/net/corda/crypto/cipher/suite/SignatureSpecEqualityTests.kt (hunk 5 above)
Comment (anien) [reply libs/crypto/cipher-suite/src/test/kotlin/net/corda/crypto/cipher/suite/SignatureSpecEqualityTests.kt]: Sorry, I think this came from some merge conflicts. This was not supposed to be changed!

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt (hunk 4 above)
Comment (anien) [reply libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt:18]: It could, but I believe it doesn't need to. Yes, if it derives from `SignatureSpecImpl`, then some duplicate implementation would be needed, which I believe is not too bad.

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+     * Obtain string containing the [signatureName].
+     *
+     * @return A string containing the signature name.
+     */
+    override fun getSignatureName(): String {
+        return this.signatureName
+    }
+
+    override fun equals(other: Any?): Boolean {
+        return this === other || other != null && other is SignatureSpec && other.signatureName == this.signatureName
```
Comment (kyriathar) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/ParameterizedSignatureSpec.kt]:
```suggestion
        return this === other || other != null && other is ParameterizedSignatureSpec && other.signatureName == this.signatureName
```

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        val SM2_SHA256 = SignatureSpecImpl("SHA256withSM2")
+
+        /**
+         * GOST3411withGOST3410 [SignatureSpec].
+         */
+        val GOST3410_GOST3411 = SignatureSpecImpl("GOST3411withGOST3410")
+    }
+
+    override fun equals(other: Any?): Boolean {
+        return this === other || other != null && other is SignatureSpec && other.signatureName.equals(this.signatureName)
```
Comment (kyriathar) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt]:
```suggestion
        return this === other || other != null && other is SignatureSpecImpl && other.signatureName == this.signatureName
```

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/CustomSignatureSpec.kt [hunk 9]
```diff
@@ -46,21 +46,21 @@ class CustomSignatureSpec(
...
-     * Obtain string containing the [signatureName].
-     *
-     * @return A string containing the signature name.
-     */
     override fun getSignatureName(): String {
         return this.signatureName
     }
+
+    override fun equals(other: Any?): Boolean {
+        return this === other || other != null && other is SignatureSpec && other.signatureName.equals(this.signatureName)
```
Comment (kyriathar) [inline libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/CustomSignatureSpec.kt]: Can we please align this `equals` same as suggestions for others.

Comment (anien) [reply libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/CustomSignatureSpec.kt]: Done.

File: simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt [hunk 10]
```diff
@@ -62,12 +65,14 @@ class TruancyResponderFlow : ResponderFlow {
...
         }
 
-        verificationService.verify(
-            serializationService.serialize(record.absentees).bytes,
-            record.signature.bytes,
-            keyOfSignature,
-            SignatureSpec.ECDSA_SHA256
-        )
+        signatureSpecService.defaultSignatureSpec(keyOfSignature)?.let {
+            verificationService.verify(
```
Comment (kyriathar) [inline simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt]: This is not right here, at signature verify site we should not getting a default signature spec. We should instead grabbing the signature spec from the signature packet. I know this was already there so maybe worth raising a JIRA to hand it over to team responsible for simulator to handle this.

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt [hunk 11]
```diff
@@ -334,14 +333,20 @@ class RpcSmokeTestFlow : ClientStartableFlow {
...
         log.info("Crypto - Signature $signedBytes received")
-        digitalSignatureVerificationService.verify(
-            bytesToSign,
-            signedBytes.bytes,
-            publicKey,
-            SignatureSpec.ECDSA_SHA256
-        )
+        signatureSpecService.defaultSignatureSpec(publicKey)?.let {
+            if (signedBytes != null) {
+                digitalSignatureVerificationService.verify(
```
Comment (kyriathar) [inline testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: Again here we should be grabbing signature spec from signature packet, and not going through `SignatureSpecService` at verify site.

File: simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt (hunk 10 above)
Comment (kyriathar) [reply simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt]: Actually, signature spec should be there in signature packet. So maybe shouldn't be too hard to fix here directly.

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt [hunk 12]
```diff
@@ -354,15 +359,21 @@ class RpcSmokeTestFlow : ClientStartableFlow {
...
         return try {
-            digitalSignatureVerificationService.verify(
-                bytesToSign,
-                signedBytes.bytes,
-                publicKey,
-                SignatureSpec.RSA_SHA256
-            )
+            signatureSpecService.defaultSignatureSpec(publicKey, DigestAlgorithmName("SHA-512"))?.let {
+                if (signedBytes != null) {
+                    digitalSignatureVerificationService.verify(
```
Comment (kyriathar) [inline testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: I think in order to achieve same behavior as it used to be, we should create an RSA key then reference it here and grab its default signature spec which would be same as it used to be (i.e. `SignatureSpec.RSA_SHA256`). But I think we should rather move this test somewhere more internally.

File: notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt [hunk 13]
```diff
@@ -54,12 +56,14 @@ fun validateRequestSignature(notarizationRequest: NotarizationRequest,
...
 
     try {
-        signatureVerifier.verify(
-            expectedSignedBytes,
-            digitalSignature.bytes,
-            requestingPartyKey,
-            SignatureSpec.ECDSA_SHA256 // TODO This shouldn't be hardcoded?
-        )
+        signatureSpecService.defaultSignatureSpec(requestingPartyKey)?.let {
+            signatureVerifier.verify(
```
Comment (kyriathar) [inline notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt]: Again here, need to grab signature spec from signature packet instead.

File: libs/crypto/cipher-suite/src/test/java/net/corda/crypto/cipher/suite/SignatureSpecEqualityJavaApiTest.java
Comment (kyriathar) [reply libs/crypto/cipher-suite/src/test/java/net/corda/crypto/cipher/suite/SignatureSpecEqualityJavaApiTest.java:1]: Actually these java API tests were moved when we internalized `cipher-suite` module. They should be deleted, but under different PR.

File: libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt (hunk 1 above)
Comment (anien) [reply libs/crypto/cipher-suite/src/main/kotlin/net/corda/crypto/cipher/suite/SignatureSpecImpl.kt]: Done. PTAL.

File: simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt (hunk 10 above)
Comment (anien) [reply simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt]: Done.

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt (hunk 11 above)
Comment (anien) [reply testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: Done.

File: simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt (hunk 10 above)
Comment (anien) [reply simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt]: I have changed this one. PTAL.

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt (hunk 11 above)
Comment (anien) [reply testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: I have changed this one, PTAL.

File: notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt (hunk 13 above)
Comment (anien) [reply notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt]: Shall we raise a ticket for this one to be fixed later?

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt (hunk 12 above)
Comment (anien) [reply testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: I would say, let's raise a ticket for this one.

File: notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt (hunk 13 above)
Comment (kyriathar) [reply notary-plugins/notary-plugin-common/src/main/kotlin/com/r3/corda/notary/plugin/common/PluggableNotaryFlowHelpers.kt]: Fixed

File: simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt (hunk 10 above)
Comment (kyriathar) [reply simulator/example-app/src/main/kotlin/net/cordacon/example/rollcall/TruancyResponderFlow.kt]: Fixed

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt (hunk 11 above)
Comment (kyriathar) [reply testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: Fixed

File: testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt (hunk 12 above)
Comment (kyriathar) [reply testing/cpbs/test-cordapp/src/main/kotlin/net/cordapp/testing/smoketests/flow/RpcSmokeTestFlow.kt]: This is now fixed with your suggestion so no need for ticket.

**Review by dickon — APPROVED** [review]
LGTM for Crypto.

**Review by blsemo — APPROVED** [review]
